packages/@aws-cdk/aws-applicationsignals-alpha/lib/slo/metric.ts

```diff
@@ -0,0 +1,357 @@
+     * @required
+     */
+    readonly keyAttributes: { [key: string]: string };
+
+    /**
+     * The name of the operation being measured
+     * Used to filter metrics for specific operation
+     *
+     */
+    readonly operationName?: string;
+
+    /**
+     * The period in seconds for metric aggregation
+     * Must be a multiple of 60
+     *
+     * @required
+     */
+    readonly periodSeconds: number;
+
+    /**
+     * The statistic to use for aggregation
+     * Examples: Average, Sum, p99
+     *
+     * @required
+     */
+    readonly statistic: MetricStatistic;
+}
+
+/**
+ * Period-based metric properties with CloudWatch metrics
+ */
+export interface PeriodBasedCloudWatchMetricProps extends SliMetricBaseProps {
+    /**
+     * The metric data queries to execute
+     * Can include raw metrics and math expressions
+     *
+     * @required
+     */
+    readonly metricDataQueries: MetricDataQuery[];
+
+    /**
+     * The period in seconds for metric aggregation
+     * Must be a multiple of 60
+     *
+     * @required
+     */
+    readonly periodSeconds: number;
+
+    /**
+     * The statistic to use for aggregation
+     * Examples: Average, Sum, p99
+     *
+     * @required
+     */
+    readonly statistic: MetricStatistic;
+}
+
+/**
+ * Request-based metric properties with Application Signals
+ */
+export interface RequestBasedAppSignalsMetricProps extends SliMetricBaseProps {
+    /**
+     * The type of metric being measured
+     * Can be LATENCY or AVAILABILITY
+     *
+     * @required
+     */
+    readonly metricType: MetricType;
+
+    /**
+     * Key attributes for the applications being monitored
+     * Must include at least one of Type, Name, and Environment
+     *
+     * @required
+     */
+    readonly keyAttributes: { [key: string]: string };
+
+    /**
+     * The name of the operation being measured
+     *
+     */
+    readonly operationName?: string;
+}
+
+/**
+ * Request-based metric properties with CloudWatch metrics
+ */
+export interface RequestBasedCloudWatchMetricProps extends SliMetricBaseProps {
+    /**
+     * Metrics that count successful requests
+     * Optional if can be derived from total - bad
+     * Used to calculate success rate
+     *
+     * @required
+     */
+    readonly goodCountMetrics: MetricDataQuery[];
+
+    /**
+     * Metrics that count total requests
+     * Used as denominator for success rate
+     *
+     * @required
+     */
+    readonly totalCountMetrics: MetricDataQuery[];
+
+    /**
+     * Metrics that count failed requests
+     * Optional if can be derived from total - good
+     *
+     */
+    readonly badCountMetrics?: MetricDataQuery[];
+}
+
+/**
+ * Period-based metric properties
+ */
+export type PeriodBasedMetricProps = PeriodBasedAppSignalsMetricProps | PeriodBasedCloudWatchMetricProps;
+
+/**
+ * Request-based metric properties
+ */
+export type RequestBasedMetricProps = RequestBasedAppSignalsMetricProps | RequestBasedCloudWatchMetricProps;
```

Comment:
We do not use union types. Use Inheritance for this pattern, also do you want that type to be exposed to customers, if no, use @internal annotation on it.